Add bit removal methods to SparseBitMatrix and factor *BitSet relational methods into more extensible trait

willcrichton opened the pull request:
I need the ability to clear the bits out of a row from `SparseBitMatrix`. Currently, all the mutating methods only allow insertion of bits, and there is no way to get access to the underlying data. 

One approach is simply to make `ensure_row` public, since it grants `&mut` access to the underlying `HybridBitSet`. This PR adds the `pub` modifier. However, presumably this method was private for a reason, so I'm open to other designs. I would prefer general mutable access to the rows, because that way I can add many mutating operations (`clear`, `intersect`, etc.) without filing a PR each time :-)

r? @ecstatic-morse

ecstatic-morse commented:
`row` is public, so the fact that each row is stored as a `HybridBitSet` is already exposed to consumers, and this change doesn't break encapsulation. However, it still feels a bit wrong to me. Calling `ensure_row` may cause the `rows` vector to grow, which isn't even necessary if you're just clearing bits.

Is there anything you could need beyond `clear`, `intersect` and `remove_bit`? That won't be too much effort, and might benefit other `rustc` users someday. I'd prefer you go that route if possible, but this is pretty low stakes.

willcrichton commented:
I could implement those methods, although they would still end up calling `ensure_row` internally. An alternative would be to add the following to parallel `SparseBitMatrix::row`:

```rust
fn row_mut(&mut self, row: R) -> Option<&mut HybridBitSet<C>>;
```

This eliminates the weird side effect, and also matches the existing API. But if you have any preference towards lifting the set methods to matrix, I can do that instead.

ecstatic-morse commented:
My whole point is that they would *not* call `ensure_mut` internally. There's no need to ensure that a row has backing storage just to intersect with or clear it.

willcrichton commented:
Ah I understand, I can do that.

One other question. For operations like `intersect` and `subtract`, would you like me to implement those methods on `SparseBitSet` / `HybridBitSet` first, and then extend them to `SparseBitMatrix`? I've already had to implement these methods in my own codebase, but they seem useful to have in rustc proper: https://github.com/willcrichton/flowistry/blob/032a5982984291b3d1e791706d0607b7be09410c/src/core/indexed.rs#L204-L234

ecstatic-morse commented:
> One other question. For operations like `intersect` and `subtract`, would you like me to implement those methods on `SparseBitSet` / `HybridBitSet` first, and then extend them to `SparseBitMatrix`? I've already had to implement these methods in my own codebase, but they seem useful to have in rustc proper: https://github.com/willcrichton/flowistry/blob/032a5982984291b3d1e791706d0607b7be09410c/src/core/indexed.rs#L204-L234

Yes, I think that's the right path, although it means handling all combinations in `{SparseBitSet, BitSet}^2`. I thought we did this already, but I guess I was wrong. This is useful, but a bit more work than I was expecting. I wonder if some parts of `rustc` convert to a dense `BitSet` just to do subtraction? 

You probably know this, but an in-tree impl should take advantage of the fact that `SparseBitSet` is in sorted order, which means you can do both `intersect` and `subtract` in a single pass with `ArrayVec::retain`.

ecstatic-morse commented:
If you don't want to make the changes to `HybridBitSet` now, I'm willing to sign off on the visibility change for `ensure_row` as long as you add some documentation and a FIXME noting that it should be removed in favor of `subtract_row`, `clear_row`, etc. when the corresponding methods are available on `HybridBitSet`.

Ideally we would make the `HybridBitSet` changes now, though. I'm happy to write up mentoring instructions, since there's a few helper traits (`SubtractFromBitSet`) that should be rewritten into more general forms.

willcrichton commented:
I'm happy to do the broader `HybridBitSet` changes. Let me know what additional changes you'd like (eg helper traits) and I can do it today / tomorrow.

ecstatic-morse commented:
Alright, so we have three different bitset types. `SparseBitSet`, `BitSet`, and `HybridBitSet` (the union of the other two). `clear`, `insert` and `remove` can be trivially implemented for all of them (and mostly already have been). We also have three operations we want to implement---intersection, union and subtraction. The current approach uses helper traits to make these operations generic over the different bitsets (intersection is missing at the moment):

https://github.com/rust-lang/rust/blob/b5fe3bc0650279e4283a0511670c0668fc0f2217/compiler/rustc_index/src/bit_set.rs#L213-L225

We want to change these to make **both** sides of the operation generic, just like (for example) [`ops::AddAssign`](https://doc.rust-lang.org/std/ops/trait.AddAssign.html). Note that the existing impls put the mutable argument on the right, unlike `AddAssign`. You can leave this as is, or change it, up to you. This will require 3^2 = 9 impls, but only 4 of those are interesting: the other 5 simply dispatch to the appropriate variant of `HybridBitSet`. The majority of the interesting ones already exist in the code; you just have to collect them in the appropriate place. Don't worry about going from a dense bitset back to a sparse one.

Finally, you should expose the operations as inherent methods on each of the bitset types, just like the current code does.

Once all that is done, it should be easy to add generic versions of `intersect_row`, `subtract_row`, and `union_row` to `SparseBitMatrix`, as well as `remove` and `clear`. You should avoid calling `ensure_row` unless its actually necessary. You'll have a small problem with the lazy, generic version of `union_row`, but you'll find a solution :smile: .

Let me know if any of that is unclear, or if you have a better approach. And if you want to just do the `ensure_row` thing so you can continue working on your project, just say the word. This was a bit more involved than I had first thought. :laughing:

willcrichton commented:
I'm trying to understand the reason for having separate a `UnionIntoBitSet::union_into` trait method and `BitSet::union` inherent method. Is it for convenience so downstream users don't have to import the `UnionIntoBitSet` trait? As opposed to just having users call `union_into` directly.

ecstatic-morse commented:
That's correct.

willcrichton commented:
Still need to add comments and tests, but @ecstatic-morse can you let me know if this general design looks right to you? In short:
* I consolidated the union/subtract (+ intersect) traits into a single `BitRelations` trait (otherwise there was an ungodly number of impl blocks).
* I added a `bit_relations_inherent_impls` macro that adds inherent implementations of each trait method to all three BitSet types.
* I factored out all the existing union/subtract/intersect logic into impls for these traits. The only impls I'm not sure about are the subtract/intersect impls for `HybridBitSet` -- I'm not sure if you want to do something especially efficient in these cases like for union.
* I added all the new methods to `SparseBitMatrix`.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3417054833) [(plain)](https://github.com/rust-lang/rust/commit/79e0a0faf93922f77f7658d2b1b95c79c9bd9608/checks/3417054833/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling memchr v2.4.1
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.49
   Compiling unwind v0.0.0 (/checkout/library/unwind)
thread 'rustc' panicked at 'called `Result::unwrap()` on an `Err` value: CapacityError: insufficient capacity', /cargo/registry/src/github.com-1ecc6299db9ec823/arrayvec-0.7.0/src/arrayvec_impl.rs:39:32

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.56.0-nightly (9a96b9f56 2021-08-24) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z macro-backtrace -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debuginfo=0 -C debug-assertions=on -C overflow-checks=off -C link-args=-Wl,-rpath,$ORIGIN/../lib -C prefer-dynamic -C llvm-args=-import-instr-limit=10 -C embed-bitcode=yes --crate-type lib
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
query stack during panic:
#0 [mir_borrowck] borrow-checking `slice::ascii::<impl at library/core/src/internal_macros.rs:98:13: 106:14>::call_mut`
#1 [analysis] running analysis passes on this crate
error: could not compile `core`
Build completed unsuccessfully in 0:04:38
```

</details>

ecstatic-morse commented on compiler/rustc_index/src/bit_set.rs:
```diff
@@ -1025,6 +1107,33 @@ impl<R: Idx, C: Idx> SparseBitMatrix<R, C> {
+    where
+        HybridBitSet<C>: BitRelations<Set>,
+    {
+        self.ensure_row(row).union(set)
```
:+1:

The other choice was to add an `IsEmpty` trait and use it to check `set` before reserving the row, but that's overkill.

ecstatic-morse commented on compiler/rustc_index/src/bit_set.rs:
```diff
@@ -208,33 +226,167 @@ impl<T: Idx> BitSet<T> {
+        // FIXME(willcrichton): should there be an optimized sparse / dense version?
+        match self {
+            HybridBitSet::Sparse(self_sparse) => self_sparse.intersect(other),
+            HybridBitSet::Dense(self_dense) => {
+                <BitSet<T> as BitRelations<HybridBitSet<T>>>::intersect(self_dense, other)
```
Is the UFCS significant here? Might be good to leave a comment if so. Collisions are why the old helper traits had different names than the inherent methods.

Regarding the FIXMEs, the only obvious potential optimization is when we're intersecting a dense bitset with a sparse one. Since we know the end result will be no bigger than the sparse bitset, we could use it to replace `self` before intersecting it with the dense bitset. You'll be fighting the borrow checker with this one, so no need to do it now if you don't want.

ecstatic-morse reviewed (commented):
Thanks @willcrichton!

I left a few notes. There's an issue I didn't foresee with `SparseBitSet::union`, which might cause problems, but for the most part this looks good.

ecstatic-morse commented on compiler/rustc_index/src/bit_set.rs:
```diff
@@ -208,33 +226,180 @@ impl<T: Idx> BitSet<T> {
+    }
+
+    fn intersect(&mut self, other: &SparseBitSet<T>) -> bool {
+        self.intersect(&other.to_dense())
```
This one is suboptimal. It would be good to leave a `FIXME` here as well.

willcrichton commented on compiler/rustc_index/src/bit_set.rs:
```diff
@@ -956,7 +1071,7 @@ impl<R: Idx, C: Idx> SparseBitMatrix<R, C> {
     }
 
-    fn ensure_row(&mut self, row: R) -> &mut HybridBitSet<C> {
+    pub fn ensure_row(&mut self, row: R) -> &mut HybridBitSet<C> {
```
Ah no, that's leftover from the initial PR

rust-timer commented:
Finished benchmarking try commit (988d819e13d73f484d8370357ba9d1f8696de90f): [comparison url](https://perf.rust-lang.org/compare.html?start=4a6547cca6d2b2f465f01331927855734687b527&end=988d819e13d73f484d8370357ba9d1f8696de90f).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up.  

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

ecstatic-morse commented:
I gave this one last look today. Given the number of bugs found during the review process, it seemed wise. Everything looks correct.

For posterity, in addition to the `SparseBitMatrix` changes, this PR extends `HybridBitSet` with in-place versions of the fundamental set operations. They aren't quite optimal, since they fail to take into account the sorted-ness of the elements in a sparse bitset, but should be no slower than the existing impls. As a result, we have the option to use `HybridBitSet` in more places (e.g. dataflow).

@bors r+

bors commented:
:pushpin: Commit e340a0e249c2efc069235c9db47e47c3e136570c has been approved by `ecstatic-morse`

<!-- @bors r=ecstatic-morse e340a0e249c2efc069235c9db47e47c3e136570c -->
<!-- homu: {"type":"Approved","sha":"e340a0e249c2efc069235c9db47e47c3e136570c","approver":"ecstatic-morse"} -->

bors commented:
:hourglass: Testing commit e340a0e249c2efc069235c9db47e47c3e136570c with merge 608b5e1c209ffb4d6d0cf83817c823b12bbb7659...
<!-- homu: {"type":"BuildStarted","head_sha":"e340a0e249c2efc069235c9db47e47c3e136570c","merge_sha":"608b5e1c209ffb4d6d0cf83817c823b12bbb7659"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3482513597)
Approved by: ecstatic-morse
Pushing 608b5e1c209ffb4d6d0cf83817c823b12bbb7659 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ecstatic-morse","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3482513597"},"merge_sha":"608b5e1c209ffb4d6d0cf83817c823b12bbb7659"} -->
